### DISCUSSION TITLE
Falang compatibility for multilingual

### DISCUSSION THREAD

**Discussion by kvintvs** [templaza/astroid-framework#697] [Q&A] [answered]
Hi, 

Currently rebuilding a site, originally made with Gantry 5, by using Astroid 3. Really impressive framework, I love it so far.

BUT... it's a quite big website translated in 4 languages. Multilingual in Joomla is implemented in a rather pecular way, as it doesn't separate layout/design from content. This means tons of duplicates if you try to maintain multiple translations, and a nightmare each time you need to do changes to your template, menu structures, etc. Fortunately, we have been using the Falang extension which requires only one instance of menus, articles, modules and Gantry particles in a source language (English in our case), and provides an interface to translate just content.

Astroid widgets and content elements don't show up in Falang. It looks like it needs to be defined as content element, like Gantry particles or Hikashop content for example. Content elements seems to be deifned in an XML file. I'm not sure who is supposed to provide that file for widgets to show up in Falang, Astroid or Falang. It looks like it's also possible to upload a content element definition XML manually in Falang, but I have no idea how to proceed. Anyone knows anything? Falang is used by lots of multilingual websites, so I'm sure this must be of interest for other Astroid users. 

Here's what Falang's help says about content elements: 

> Content Elements are the definition files that map a certain table in your database to your translations. Within these elements the detail configuration of the mapping is defined.
> 
> As config elements are somehow complex definitions and you can mess quite a lot with wrong configuration it is recommended to ask your extension developer to provide you the correct XML files. The creation of these files is simple and should be easy for any developer. 

[Here's the documentation I could find](https://www.faboba.com/en/composants/falang/documentation/55-creation-of-content-elements.html) about content element definitions on Falang's website.

Anyone has had any luck translating Astroid widgets with Falang?

**Comment by Chacapamac** [comment]
@kvintvs 

Falang, Gantry ?????, sorry to completely disagree with you.
I hate both, with a passion, no disrespect.

I worked on Joomla from the 1.5 version when Joomfish third-party component was the only viable solution to add the multilanguage ability to Joomla. Rockeththeme was my Framework until the apparition of the “new” Gantry and the slow descent in a proprietary framework that imposed its own “System” over the great Joomla, trying to reinvent the wheel over this now, year after year, recognized “Best Free CMS” and ”Best Open Source CMS” in the World by the Prestigious “CMS Critic People's Choice Awards” since 2011 (https://docs.joomla.org/Joomla!_Awards). 

I live in Quebec and ALL my websites were multilingual (I’m 65 now). I can say from experience that the actual Joomla multilingual system is a fantastic asset to this great CMS. I have created large community websites with tons of content and I can tell you, that this system makes managing multilingual content a cinch, and that, without adding any third-party component that can disappear at any moment, add problems and become unnecessary crutches that cripple Joomla more than help.

When I try to replace my trusted Joomla and sadly metamorphose Rockettheme Gantry framework I embark on a year-long evaluation of ALL Joomla frameworks available at that time. 

One of my major criteria was to choose a framework that uses 100% of Joomla's core ability. After all the testing, Joomdev Astroid was my final winner, and by far the best Framework and still is today, because it REALLY uses the power of Joomla and even enhances it and continues to be the perfect companion to Joomla. Hopefully, it will stay like that.

**Reply by kvintvs** [reply]
> One of my major criteria was to choose a framework that uses 100% of Joomla's core ability. After all the testing, Joomdev Astroid was my final winner, and by far the best Framework and still is today, because it REALLY uses the power of Joomla and even enhances it and continues to be the perfect companion to Joomla. Hopefully, it will stay like that.

Yes, this is exactly what I have been through, and concluded that Astroid was indeed my best option. My aim is to use as many core features as possible, because it's a headache to update the site with many 3rd party extensions, but for the multilingual part it's not yet reasonable to drop Falang. 

Once you start maintaining a site translated to more than 1 language, you quickly realize the amount of redundancies that Joomla multilingual system imposes. You need to duplicate menus, templates, articles, modules, categories, etc. for each language. I don't understand why. Those are supposed to be _translations_ of a source content, not "similar" items. I read another post about multilingual in here where a user had mutliple templates for each language, having to maintain dozens of templates in total. If that user wants to change a simple colour nuance on his site, he's in for hours of clicking, scrolling and and copy/pasting. It doesn't make sense IMO.

I think it's great if Joomla's handling of multilingual sites suits you needs, but it doesn't work for me and I hope you will reckon that it might not be ideal for users like me with many translations to maintain. In my opinion, you should be able to build and maintain the site in _one source language_, which can then be translated to as many target languages you want. This is how Drupal's core multilingual feature works for example. This is how mutiple web and native apps I have worked on are strucutred as well.

Has anyone in here successfully translated Astroid widgets with Falang?

**Comment by Chacapamac** [comment]
I agree that Astroid can probably be compatible with Falang, but not to the detriment of its ability to work with Joomla's core multilingual system.

**Reply by kvintvs** [reply]
I don't think Falang compatibility requres any changes to Astroid, other than providing a definition of its content in an XML file (see my first post for link to documentation)

**Comment by Chacapamac** [comment]
I think @sonvnn just can take a look with you to see how to make Astroid compatible with Falang. If it’s relatively easy to do and do not change the ability of Astroid to use the Joomla core multilingual system. Why Not!

**Comment by kvintvs** [comment]
That would be wonderful! As far as I can see, the only issue is that the Astroid Widgets are not showing up as content elements in Falang. Since they are not a core joomla content element, they need to be defined somehow so that Falang sees them, but I'm not sure how this done.

**Comment by sonvnn** [comment]
Hello @Chacapamac @kvintvs ,

Unfortunately. I don't have much experience with Falang. It would be better if someone with more experience with Falang could share how to integrate it with other Extensions. If you can get the author of Falang to join that would be the best thing at this time. Sorry I'm too busy right now to spend time researching Falang.

This time is tough right now, and we're also considering spending more time on projects that can generate revenue to support our team. I am sorry about that.

Best Regards,
Sonny

**Comment by sbouey** [comment]
Hi,
I'm stéphane from Falang if someone has a demo site with Joomal 5/Astoid configured with content i can have a look to see , if it's possible to support this framework
You can contact me for this on my contact page on faboba.com

Stéphane

**Reply by sonvnn** [reply]
You can download Astroid quickstart here for testing. https://github.com/templaza/astroid-framework/releases/download/v3.0.15/astroid-quickstart.zip

Please let me know if you have anyquestions. 

Thanks & Best regards,
Sonny

**Comment by sbouey** [comment]
Hi,
Ok i have installed the quickstart, it can't be translated directly with falang because the layout of page are not in module but directly in the template style, i haven't found table where the change are stored  in the layout

I think but i'm not sure yet , the way to translate astroid will be with an extra plugin , if it's possible to add dynamicly tab here
https://snipboard.io/fr7q3C.jpg
i have done this for yootheme / sp page builder on Joomla 
see here 
[https://www.faboba.com/en/components/translate-for-yootheme-builder/presentation.html](https://www.faboba.com/en/components/translate-for-yootheme-builder/presentation.html)

it's an extra tab where the translation can be add in the elements from here
\www\libraries\astroid\framework\elements

Stéphane

**Comment by sbouey** [comment]
i have look a little deeper and everything is hardoded but the idee is this
https://snipboard.io/1rUzIq.jpg
i have modified the headling.xml but in the file
libraries\astroid\framework\library\astroid\Helper.php
line 362 with an event launch before the return 
or something in the file
libraries\astroid\framework\library\astroid\Element.php
to override the xml or add dynamically fieldset it can be possible (for backend to add translation)

Stéphane

**Reply by sonvnn** [reply]
Thanks for your time in Astroid. I will be working on a Falang integration solution after the 3.1.0 release. I will contact you if I need any help.

Best Regards,
Sonny

**Comment by sbouey** [comment, marked as answer]
Hi,

It's not really a Falang integration, it's an extra plugin who can work with or without the native multlanguage system of Falang.
You can contact me directly.
I think it will not be to complicate to do.

Stéphane

**Reply by sonvnn** [reply]
Thanks for get in touch. I will contact you when I have question about Falang.

Best Regards,
Sonny

**Comment by kvintvs** [comment]
Hi @sonvnn 

Thank you for the latest updates to Astroid!
Have you by any chance had opportunity to look further into enabling Falang in Astroid?

**Reply by sonvnn** [reply]
It is very complicated at the moment. So I need more time to research. Please wait for the update :)

**Reply by kvintvs** [reply]
Ok, thank you
